drivers/cleanup.py

```diff
@@ -1501,6 +1503,9 @@ def findLeafCoalesceable(self):
             if vdi.getConfig(vdi.DB_ONBOOT) == vdi.ONBOOT_RESET:
                 Util.log("Skipping reset-on-boot %s" % vdi)
                 continue
+            if vdi.getConfig(vdi.DB_CACHING) == "true":
```

Comment:
From discussion via email, this should probably check the `allowed-caching=true` property of the VDI and block leaf-coalesce of any VHD declared to having caching enabled. As caching will be disabled in either of these situations

- The VDI has no parent
- The VDI's parent also has a parent (i.e. the tree is taller than 2 nodes)

(See - https://github.com/xapi-project/sm/blob/master/drivers/blktap2.py#L1828)

Coalescing a VDI with `allow-caching = true` will violate these constraints.

Reply:
As per discussion by email, I'm not totally convinced which approach is better, but I've created another version for review & consideration (PR #385).